src/Oscoin/Data/Core.hs

```diff
@@ -0,0 +1,495 @@
+    }
+
+defaultProject :: Address c -> Project c
+defaultProject addr = Project
+    { pAccount          = mkAccount addr
+    , pMaintainers      = mempty
+    , pContributors     = mempty
+    , pSupporters       = mempty
+    , pDependencies     = mempty
+    , pCheckpoints      = mempty
+    , pSendTransfer     = defaultSendTransfer
+    , pReceiveTransfer  = defaultReceiveTransfer
+    , pReceiveReward    = defaultReceiveReward
+    , pCheckpoint       = defaultCheckpoint
+    , pUnregister       = defaultUnregister
+    , pAuthorize        = defaultAuthorize
+    , pDeauthorize      = defaultDeauthorize
+    , pUpdateContract   = defaultUpdateContract
+    }
+
+-- | Get the address of a project.
+projectAddr :: Project c -> Address c
+projectAddr = accountAddr . pAccount
+
+-- | A delegation of oscoin between two accounts. Allows members to become
+-- *supporters* by delegating their voting rights to a project. Allows projects
+-- to use the delegated tokens to vote.
+data Delegation c = Delegation
+    { dDelegator  :: Address c
+    -- ^ Account delegating.
+    , dReceiver   :: Address c
+    -- ^ Account receiving the delegation.
+    , dBalance    :: Balance
+    -- ^ Balance being delegated.
+    , dCommitment :: Height
+    -- ^ Time commitment of the delegation.
+    , dSince      :: Height
+    -- ^ Starting time of the delegation.
+    }
+
+-- | A member of a project. Includes maintainers, contributors and supporters.
+data Member c = Member
+    { memberAccount       :: Address c
+    , memberDelegation    :: Maybe (Delegation c)
+    , memberSince         :: Epoch
+    , memberContributions :: Set (Contribution c)
+    }
+
+mkMember
+    :: (Ord (Signature c), Ord (BlockHash c), Ord (PublicKey c))
+    => Address c -> Epoch -> Member c
+mkMember addr e = Member
+    { memberAccount       = addr
+    , memberDelegation    = Nothing
+    , memberSince         = e
+    , memberContributions = mempty
+    }
+
+-- | A dependency between two projects.
+data Dependency c = Dependency
+    { depFrom :: Address c
+    , depTo   :: Address c
+    , depHash :: Crypto.Hash c
+    -- ^ Hash of the checkpoint state being depended on.
+    } deriving (Generic)
+
+-- | Identifies a contract handler.
+data Handler =
+      SendTransferHandler
+    | ReceiveRewardHandler
+    | ReceiveTransferHandler
+    | CheckpointHandler
+    | AuthorizeHandler
+    | DeauthorizeHandler
+    | UnregisterHandler
+    | UpdateContractHandler
+    deriving (Show, Eq, Ord)
+
+-- | Dictionary used to parametrize a handler.
+type HandlerParams = Map ParamKey ParamVal
+
+-- | Parameter key.
+type ParamKey = Text
+
+-- | Parameter value.
+data ParamVal =
+      Integer  Int
+    -- ^ Expresses positive or negative whole numbers.
+    | Rational Int Int
+    -- ^ Expresses ratios, eg. 1/2.
+    deriving (Show, Eq, Ord)
+
+-------------------------------------------------------------------------------
+
+newtype Error = Error Text
```

Comment:
I think this name is too generic, especially if this module is imported unqualified. What do you think about `ContractError` or `HandlerError`